### PR TITLE
1.2 JS clean-up

### PR DESCRIPTION
**Pull request by davwheat**
**Changes proposed in this pull request:**

Performs some clean-up to frontend JS, ready for 1.2.

- fixes `extend.ts` typing issue introduced in #2928 
- fixes bad JS docblocks (contained within f476c2b9821e0cd9b49cf798bf6582fe98f74477)
- remove fallbacks for `window.requestAnimationFrame()` -- this is supported in all browsers now
- consistently return `null` instead of `''` or `[]` for empty vdoms
- simplify some code and checks inside `Post`, `PostsUserPage` and `slideable` (40be94daaa08ee504b278f09023a25e1759a594c)

> Updating our JSDoc blocks allows us to have _some_ form of type safety and hinting without needing to fully convert to Typescript just yet.

**Reviewers should focus on:**

N/A

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

File: js/src/forum/components/Post.js [hunk 1]
```diff
@@ -133,11 +133,11 @@ export default class Post extends Component {
       classes.push('Post--loading');
     }
 
-    if (user && user === app.session.user) {
+    if (user === app.session.user) {
```
Comment (askvortsov1) [inline js/src/forum/components/Post.js:136]: Doesn't this prevent the class being applied on posts with deleted users while browsing as guest?

File: js/src/forum/utils/slidable.js [hunk 2]
```diff
@@ -36,7 +36,7 @@ export default function slidable(element) {
     // bit of a workaround. We set up the animation with a step function that
     // will set the transform property, but then we animate an unused property
     // (background-position-x) with jQuery.
-    options.duration = options.duration || 'fast';
+    options.duration ||= 200;
```
Comment (askvortsov1) [inline js/src/forum/utils/slidable.js:39]: Do we need to change it to a magic number? If so, let's put it in a constant

File: js/src/admin/components/DashboardWidget.js [hunk 3]
```diff
@@ -20,6 +20,6 @@ export default class DashboardWidget extends Component {
    * @return {import('mithril').Children}
    */
   content() {
-    return [];
+    return null;
   }
```
Comment (askvortsov1) [inline js/src/admin/components/DashboardWidget.js:24]: These should really all become abstract methods for v2

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

File: js/src/forum/utils/slidable.js (hunk 2 above)
Comment (davwheat) [reply js/src/forum/utils/slidable.js:39]: It's just as magic as the word `fast`, which acts as a synonym for `200` internally within jQuery.

File: js/src/forum/components/Post.js (hunk 1 above)
Comment (davwheat) [reply js/src/forum/components/Post.js:136]: Wow, I would have never thought of that...
```suggestion
    if (user && user === app.session.user) {
```

File: js/src/common/extend.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@
  * @param methods The name or names of the method(s) to extend
  * @param callback A callback which mutates the method's output
  */
-export function extend<T extends object, K extends KeyOfType<T, Function>>(
+export function extend<T extends Record<any, any>, K extends KeyOfType<T, Function>>(
```
[the remark this answers is not in the document]
Comment (davwheat) [reply js/src/common/extend.ts:26]: I thought any was safer, but in reality, it'll always be string as far as I can think.

```suggestion
export function extend<T extends Record<string, any>, K extends KeyOfType<T, Function>>(
```

File: js/src/admin/components/BasicsPage.js [hunk 5]
```diff
@@ -120,8 +120,7 @@ export default class BasicsPage extends AdminPage {
    * Build a list of options for the default homepage. Each option must be an
    * object with `path` and `label` properties.
    *
-   * @return {ItemList}
-   * @public
+   * @return {ItemList<{ path: string, label: import('@askvortsov/rich-icu-message-formatter').NestedStringArray }>}
```
[the remark this answers is not in the document]
Comment (davwheat) [reply js/src/admin/components/BasicsPage.js:123]:
```suggestion
   * @return {ItemList<{ path: string, label: import('mithril').Children }>}
```

File: js/src/forum/components/NotificationGrid.js [hunk 6]
```diff
@@ -19,21 +19,21 @@ export default class NotificationGrid extends Component {
     /**
      * Information about the available notification methods.
      *
-     * @type {Array}
+     * @type {({ name: string, icon: string, label: import('@askvortsov/rich-icu-message-formatter').NestedStringArray })[]}
```
[the remark this answers is not in the document]
Comment (davwheat) [reply js/src/forum/components/NotificationGrid.js:22]:
```suggestion
     * @type {({ name: string, icon: string, label: import('mithril').Children })[]}
```

File: js/src/forum/utils/slidable.js (hunk 2 above)
Comment (askvortsov1) [reply js/src/forum/utils/slidable.js:39]: Sure, but the string `fast` is descriptive and feels less magic than `200`.

Comment (davwheat) [reply js/src/forum/utils/slidable.js:39]:
```suggestion
    options.duration ||= 'fast';
```

File: js/src/forum/components/EventPost.js [hunk 7]
```diff
@@ -54,8 +54,8 @@ export default class EventPost extends Post {
   /**
    * Get the description text for the event.
    *
-   * @param {Object} data
-   * @return {String|Object} The description to render in the DOM
+   * @param {Record<string, unknown>} data
+   * @return {import('@askvortsov/rich-icu-message-formatter').NestedStringArray} The description to render in the DOM
```
Comment (askvortsov1) [inline js/src/forum/components/EventPost.js:58]: This still needs to be Mithril.Children

File: js/src/forum/components/NotificationGrid.js [hunk 8]
```diff
@@ -19,21 +19,21 @@ export default class NotificationGrid extends Component {
...
-     * @type {Object}
+     * @type {Record<string, boolean>}
      */
     this.loading = {};
 
     /**
      * Information about the available notification types.
      *
-     * @type {Array}
+     * @type {({ name: string, icon: string, label: import('@askvortsov/rich-icu-message-formatter').NestedStringArray })[]}
```
Comment (askvortsov1) [inline js/src/forum/components/NotificationGrid.js:36]: Same here, also should be Mithril.Children

File: js/src/forum/components/NotificationGrid.js [hunk 9]
```diff
@@ -168,7 +168,7 @@ export default class NotificationGrid extends Component {
    * - `icon` The icon to display in the column header.
    * - `label` The label to display in the column header.
    *
-   * @return {ItemList}
+   * @return {ItemList<{ name: string, icon: string, label: import('@askvortsov/rich-icu-message-formatter').NestedStringArray }>}
```
Comment (askvortsov1) [inline js/src/forum/components/NotificationGrid.js:171]: And here

File: js/src/forum/components/NotificationGrid.js [hunk 10]
```diff
@@ -197,7 +197,7 @@ export default class NotificationGrid extends Component {
    * - `icon` The icon to display in the notification grid row.
    * - `label` The label to display in the notification grid row.
    *
-   * @return {ItemList}
+   * @return {ItemList<{ name: string, icon: string, label: import('@askvortsov/rich-icu-message-formatter').NestedStringArray }>}
```
Comment (askvortsov1) [inline js/src/forum/components/NotificationGrid.js:200]: And here

**Review by askvortsov1 — APPROVED** [review]
Approving as long as the last few `NestedStringArray` are changed to `Mithril.Children`

File: js/src/forum/components/EventPost.js (hunk 7 above)
Comment (davwheat) [reply js/src/forum/components/EventPost.js:58]:
```suggestion
   * @return {import('mithril').Children} The description to render in the DOM
```

File: js/src/forum/components/NotificationGrid.js (hunk 8 above)
Comment (davwheat) [reply js/src/forum/components/NotificationGrid.js:36]:
```suggestion
     * @type {({ name: string, icon: string, label: import('mithril').Children })[]}
```

File: js/src/forum/components/NotificationGrid.js (hunk 9 above)
Comment (davwheat) [reply js/src/forum/components/NotificationGrid.js:171]:
```suggestion
   * @return {ItemList<{ name: string, icon: string, label: import('mithril').Children }>}
```

File: js/src/forum/components/NotificationGrid.js (hunk 10 above)
Comment (davwheat) [reply js/src/forum/components/NotificationGrid.js:200]:
```suggestion
   * @return {ItemList<{ name: string, icon: string, label: import('mithril').Children}>}
```

**Review by datitisev — APPROVED** [review]
Not tested locally (shouldn't need to) & changes are mostly docblocks which don't break anything. Nothing stands out to me so LGTM.

**Pull request merged**